pandas/tests/window/test_rolling.py

```diff
@@ -868,3 +868,51 @@ def test_rolling_period_index(index, window, func, values):
     result = getattr(ds.rolling(window, closed="left"), func)()
     expected = pd.Series(values, index=index)
     tm.assert_series_equal(result, expected)
+
+
+@pytest.mark.parametrize("method", ["var", "mean", "sum", "skew", "kurt"])
+def test_rolling_variable_non_monotonic(method):
+    """
+    Make sure the (rare) branch of variable window-size with non-monotonic indices is
+    covered by a test.
+
+    GH 36933.
+    """
+    # output from 1.1.3 is assumed to be the expected output
+    nan = float("NaN")
+    expected = {
+        "sum": [nan, 10.0, nan, 7.0, 10.0],
+        "mean": [nan, 2.5, nan, 1.75, 2.0],
+        "var": [nan, 1.6666666666666667, nan, 4.249999999999999, 2.5],
+        "skew": [nan, 0.0, nan, 0.19973522061815654, 0.0],
+        "kurt": [
+            nan,
+            -1.1999999999999993,
+            nan,
+            -4.858131487889274,
+            -1.2000000000000028,
+        ],
+    }
+
+    # Example found in computation.rst
+    use_expanding = [True, False, True, False, True]
+    df = pd.DataFrame({"values": range(5)})
+
+    class CustomIndexer(pd.api.indexers.BaseIndexer):
+        def get_window_bounds(self, num_values, min_periods, center, closed):
+            start = np.empty(num_values, dtype=np.int64)
+            end = np.empty(num_values, dtype=np.int64)
+            for i in range(num_values):
+                if self.use_expanding[i]:
+                    start[i] = 0
+                    end[i] = i + 1
+                else:
+                    start[i] = i
+                    end[i] = i + self.window_size
```

Comment:
I think this may cause the randomness! `i + self.window_size` could potentially be larger than the entire array leading to accessing random data (or segfaults?)

Reply:
I changed the test to make sure that it cannot exceed the array size. Let's see what happens :)

I honestly do not know how `BaseIndex` and `rolling` interact but I'm surprised that we get NaNs for when not expanding (at least sum and mean should be able to calculate the sum/mean of one number?)

Reply:
forget my last comment, NaN is expected since the window size is four: "Minimum number of observations in window required to have a value (otherwise result is NA). For a window that is specified by an offset, min_periods will default to 1. Otherwise, min_periods will default to the size of the window."